app/models/user.rb

```diff
@@ -258,8 +258,14 @@ def path
   end
 
   def followed_articles
-    Article
-      .cached_tagged_with_any(cached_followed_tag_names).unscope(:select)
+    relation = Article
```

Comment:
I tried to run the generate query in Blazer and it timed out. That said, the original query without exclusions also timed out. We're only using this method in `EmailDigestArticleCollector` so I hope that's ok.

@djuber Maybe you have a good way of figuring out if this is reasonably safe to deploy?

Reply:
with no followed users (just tag follows and anti-follows) I get locally a query that looks like this

```sql
SELECT "articles".*
FROM (
  (
  -- following tags and antifollowing tags
    SELECT "articles".*
      FROM "articles"
      WHERE
        NOT ((cached_tag_list ~ '[[:<:]]computerscience[[:>:]]'))
        AND (cached_tag_list ~ '[[:<:]]go[[:>:]]' OR cached_tag_list ~ '[[:<:]]python[[:>:]]')
  )
  UNION (
    -- no following user ids, empty union query:
    SELECT "articles".* FROM "articles" WHERE 1=0
    )
) "articles";
```

Running explain on this in blazer gives the following (local explain would be biased by factors like table size and statistics and could be misleading):

```
Unique  (cost=173963.32..175036.53 rows=29403 width=1133)
  ->  Sort  (cost=173963.32..173978.02 rows=29403 width=1133)
        Sort Key: articles.id, articles.user_id, articles.title, articles.body_html, articles.slug, articles.created_at, articles.updated_at, articles.main_image, articles.description, articles.published, articles.published_at, articles.featured, articles.processed_html, articles.featured_number, articles.social_image, articles.body_markdown, articles.organization_id, articles.canonical_url, articles.collection_id, articles.show_comments, articles.main_image_background_hex_color, articles.receive_notifications, articles.approved, articles.published_from_feed, articles.comments_count, articles.reactions_count, articles.video, articles.video_code, articles.video_source_url, articles.video_thumbnail_url, articles.video_closed_caption_track_url, articles.hotness_score, articles.feed_source_url, articles.positive_reactions_count, articles.edited_at, articles.crossposted_at, articles.spaminess_rating, articles.cached_tag_list, articles.path, articles.cached_user_name, articles.cached_user_username, articles.last_comment_at, articles.email_digest_eligible, articles.boost_states, articles.video_state, articles.page_views_count, articles.previous_positive_reactions_count, articles.score, articles.reading_time, articles.comment_template, articles.video_duration_in_seconds, articles.experience_level_rating, articles.experience_level_rating_distribution, articles.last_experience_level_rating_at, articles.rating_votes_count, articles.originally_published_at, articles.organic_page_views_count, articles.organic_page_views_past_month_count, articles.organic_page_views_past_week_count, articles.cached_user, articles.cached_organization, articles.archived, articles.any_comments_hidden, articles.nth_published_by_author, articles.comment_score, articles.search_optimized_title_preamble, articles.search_optimized_description_replacement, articles.public_reactions_count, articles.previous_public_reactions_count, articles.user_subscriptions_count, articles.co_author_ids, articles.reading_list_document
        ->  Append  (cost=0.00..173526.87 rows=29403 width=1133)
              ->  Seq Scan on articles  (cost=0.00..173394.56 rows=29402 width=2290)
                    Filter: (((cached_tag_list)::text !~ '[[:<:]]computerscience[[:>:]]'::text) AND (((cached_tag_list)::text ~ '[[:<:]]go[[:>:]]'::text) OR ((cached_tag_list)::text ~ '[[:<:]]python[[:>:]]'::text)))
              ->  Result  (cost=0.00..0.00 rows=0 width=0)
                    One-Time Filter: false
```

I think the union between followed tags and followed users is imposing the sort/sort-key/append actions in the query plan?

Limiting the columns to just id and title, and limiting the result set to 10 (on the outer query, not the inner query) and re-running, I'm seeing about 3-4s in blazer to run the query, it's probably not something we want to do often or in a loop over users, and definitely not something we'd want in the frontend.

Removing the "any articles for followed users" empty union select I get a much more reasonable result

```sql
explain SELECT "articles".*
FROM (
  (
  -- following tags and antifollowing tags
    SELECT "articles".*
      FROM "articles"
      WHERE
        NOT ((cached_tag_list ~ '[[:<:]]computerscience[[:>:]]'))
        AND (cached_tag_list ~ '[[:<:]]go[[:>:]]' OR cached_tag_list ~ '[[:<:]]python[[:>:]]')
  )
-- UNION for followed users was here
) "articles" limit 10;
```

```
Limit  (cost=0.00..58.97 rows=10 width=2290)
  ->  Seq Scan on articles  (cost=0.00..173394.56 rows=29402 width=2290)
        Filter: (((cached_tag_list)::text !~ '[[:<:]]computerscience[[:>:]]'::text) AND (((cached_tag_list)::text ~ '[[:<:]]go[[:>:]]'::text) OR ((cached_tag_list)::text ~ '[[:<:]]python[[:>:]]'::text)))
```

Without the users union this query runs in 288ms in blazer (using the simpler query plan):

```sql
SELECT "articles".id, "articles".title FROM ( (SELECT "articles".* FROM "articles" WHERE NOT ((cached_tag_list ~ '[[:<:]]computerscience[[:>:]]')) AND (cached_tag_list ~ '[[:<:]]go[[:>:]]' OR cached_tag_list ~ '[[:<:]]python[[:>:]]'))  ) "articles" limit 10;
```

Reply:
Thanks for all the details @djiber, much appreciated. Also, great idea re the warning, I'll definitely add that.